v2/yarpcgrpc/dialer.go

```diff
@@ -22,32 +22,124 @@ package grpc
+	// explicitly here in case these change in grpc-go so that YARPC stays
+	// consistent.
+	defaultClientMaxRecvMsgSize = 1024 * 1024 * 4
+	defaultClientMaxSendMsgSize = math.MaxInt32
 )
 
 var _ yarpc.Dialer = (*Dialer)(nil)
 
-// Dialer is a decorator for a gRPC transport that threads dial options for
-// every retained peer.
+// Dialer keeps track of all gRPC peers
 type Dialer struct {
+	// ClientMaxRecvMsgSize is the maximum message size the client can receive.
+	//
+	// The default is 4MB.
+	ClientMaxRecvMsgSize int
+	// ClientMaxSendMsgSize is the maximum message size the client can send.
+	//
+	// The default is unlimited.
+	ClientMaxSendMsgSize int
+
+	// BackoffStrategy specifies the backoff strategy for delays between
+	// connection attempts for each peer.
+	//
+	// The default is exponential backoff starting with 10ms fully jittered,
+	// doubling each attempt, with a maximum interval of 30s. //TODO(apeatsbond)
+	Backoff yarpcconfig.Backoff
+
+	// Credentials specifies connection level security credentials (e.g.,
+	// TLS/SSL) for outbound connections.
+	Credentials credentials.TransportCredentials
+
+	// Tracer configures a logger for the dialer.
+	Logger *zap.Logger
+
+	// Tracer configures a tracer for the dialer.
+	Tracer opentracing.Tracer
+
+	internal *dialerInternals
+}
+
+type dialerInternals struct {
 	lock          sync.Mutex
-	options       *dialOptions
 	addressToPeer map[string]*grpcPeer
+
+	dialOptions []grpc.DialOption
+	backoff     backoffapi.Strategy
+
+	logger *zap.Logger
+	tracer opentracing.Tracer
 }
 
 // Start starts the gRPC dialer.
 func (d *Dialer) Start(_ context.Context) error {
-	d.addressToPeer = make(map[string]*grpcPeer)
-	// init options?
+	d.internal = &dialerInternals{
+		addressToPeer: make(map[string]*grpcPeer),
+		backoff:       backoff.DefaultExponential,
+		tracer:        opentracing.GlobalTracer(),
+		logger:        zap.NewNop(),
+	}
+
+	if d.Logger != nil {
+		d.internal.logger = d.Logger
+	}
+	if d.Tracer != nil {
+		d.internal.tracer = d.Tracer
+	}
+
+	var credentialDialOption grpc.DialOption
+	if d.Credentials == nil {
+		credentialDialOption = grpc.WithInsecure()
+	} else {
+		credentialDialOption = grpc.WithTransportCredentials(d.Credentials)
+	}
+
+	defaultCallOptions := []grpc.CallOption{grpc.CallCustomCodec(customCodec{})}
+	if d.ClientMaxRecvMsgSize != 0 {
+		defaultCallOptions = append(defaultCallOptions, grpc.MaxCallRecvMsgSize(d.ClientMaxRecvMsgSize))
+	} else {
+		defaultCallOptions = append(defaultCallOptions, grpc.MaxCallRecvMsgSize(defaultClientMaxRecvMsgSize))
+	}
+	if d.ClientMaxSendMsgSize != 0 {
+		defaultCallOptions = append(defaultCallOptions, grpc.MaxCallSendMsgSize(d.ClientMaxSendMsgSize))
+	} else {
+		defaultCallOptions = append(defaultCallOptions, grpc.MaxCallSendMsgSize(defaultClientMaxSendMsgSize))
+	}
+
+	d.internal.dialOptions = []grpc.DialOption{
+		grpc.WithUserAgent(UserAgent),
+		grpc.WithDefaultCallOptions(defaultCallOptions...),
+		credentialDialOption,
+	}
```

Comment:
nit: If we see the number of these options growing over time, it might help to have a helper function for configuring the `[]grpc.DialOption`, such as a `getDialOptions` or `setDialOptions`.

Reply:
yeah agreed, I'll move them to a helper function. I think we might end up exposing more and more gRPC options in the future